redbot/core/utils/__init__.py

```diff
@@ -255,3 +256,54 @@ def bounded_gather(
     tasks = (_sem_wrapper(semaphore, task) for task in coros_or_futures)
 
     return asyncio.gather(*tasks, return_exceptions=return_exceptions)
+
+
+class AsyncGen(AsyncIterable[_T]):
+    """Yield entry every `delay` seconds per `steps`."""
+
+    def __init__(
+        self, contents: Union[Iterator[_T], Iterable[_T]], delay: float = 0.0, steps: int = 1
```

Comment:
Considering we're passing iterable (iterator is an iterable), I think it would be clearer to call it `iterable` instead of `contents`

I also have two suggestions for typings here:
a) we can simplify `Union[Iterator[_T], Iterable[_T]]` to just `Iterable[_T]` as all iterators are iterable
b) since `float` is not an `int`, we need to specify union of both here if we want to accept both (and since it's only used for `asyncio.sleep()`, I think we do want it)
```suggestion
        self, iterable: Iterable[_T], delay: Union[float, int] = 0, steps: int = 1
```